File: go/parser/parser.go
package parser

import (
	"fmt"
	"gorb/ast"
	"gorb/lexer"
	"gorb/token"
	"strconv"
)

func New(l *lexer.Lexer) *Parser {
	p := &Parser{
		l:      l,
		errors: []string{},
	}

	p.unaryParseFns = make(map[token.TokenType]unaryParseFn)
	p.registerUnary(token.PAREN_OPEN, p.parseGroupedExpression)
	p.registerUnary(token.IF, p.parseIfExpression)
	p.registerUnary(token.BANG, p.parseUnaryExpression)
	p.registerUnary(token.MINUS, p.parseUnaryExpression)
	p.registerUnary(token.FUNCTION, p.parseFunctionLiteral)
	p.registerUnary(token.IDENTIFIER, p.parseIdentifier)
	p.registerUnary(token.TRUE, p.parseBoolean)
	p.registerUnary(token.FALSE, p.parseBoolean)
	p.registerUnary(token.INTEGER, p.parseIntegerLiteral)
	p.registerUnary(token.STRING, p.parseStringLiteral)
	p.registerUnary(token.BRACKET_OPEN, p.parseArrayLiteral)
	p.registerUnary(token.BRACE_OPEN, p.parseHashLiteral)

	p.binaryParseFns = make(map[token.TokenType]binaryParseFn)
	p.registerBinary(token.BRACKET_OPEN, p.parseIndexExpression)
	p.registerBinary(token.PAREN_OPEN, p.parseCallExpression)
	p.registerBinary(token.PLUS, p.parseBinaryExpression)
	p.registerBinary(token.MINUS, p.parseBinaryExpression)
	p.registerBinary(token.SLASH, p.parseBinaryExpression)
	p.registerBinary(token.ASTERISK, p.parseBinaryExpression)
	p.registerBinary(token.EQUALS, p.parseBinaryExpression)
	p.registerBinary(token.NOT_EQUALS, p.parseBinaryExpression)
	p.registerBinary(token.LESS_THAN, p.parseBinaryExpression)
	p.registerBinary(token.GREATER_THAN, p.parseBinaryExpression)

	p.nextToken()
	p.nextToken()

	return p
}

type Parser struct {
	l      *lexer.Lexer
	errors []string

	curToken  token.Token
	peekToken token.Token

	unaryParseFns  map[token.TokenType]unaryParseFn
	binaryParseFns map[token.TokenType]binaryParseFn
}

type (
	unaryParseFn  func() ast.Expression
	binaryParseFn func(ast.Expression) ast.Expression
)

func (p *Parser) registerUnary(tt token.TokenType, fn unaryParseFn) {
	p.unaryParseFns[tt] = fn
}

func (p *Parser) registerBinary(tt token.TokenType, fn binaryParseFn) {
	p.binaryParseFns[tt] = fn
}

func (p *Parser) nextToken() {
	p.curToken = p.peekToken
	p.peekToken = p.l.NextToken()
}

func (p *Parser) ParseProgram() *ast.Program {
	program := &ast.Program{}
	program.Statements = []ast.Statement{}

	for !p.curTokenIs(token.EOF) {
		stmt := p.parseStatement()
		if stmt != nil {
			program.Statements = append(program.Statements, stmt)
		}
		p.nextToken()
	}

	return program
}

// statements

func (p *Parser) parseStatement() ast.Statement {
	switch p.curToken.Type {
	case token.RETURN:
		return p.parseReturnStatement()
	case token.DECLARATION:
		return p.parseDeclarationStatement()
	default:
		return p.parseExpressionStatement()
	}
}

func (p *Parser) parseReturnStatement() *ast.ReturnStatement {
	stmt := &ast.ReturnStatement{Token: p.curToken}

	p.nextToken()

	stmt.ReturnValue = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.SEMICOLON) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseDeclarationStatement() *ast.DeclarationStatement {
	stmt := &ast.DeclarationStatement{Token: p.curToken}

	if !p.expectPeek(token.IDENTIFIER) {
		return nil
	}

	stmt.Name = &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}

	if !p.expectPeek(token.ASSIGN) {
		return nil
	}

	p.nextToken()

	stmt.Value = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.SEMICOLON) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseExpressionStatement() *ast.ExpressionStatement {
	stmt := &ast.ExpressionStatement{Token: p.curToken}

	stmt.Expression = p.parseExpression(LOWEST)

	for p.peekTokenIs(token.SEMICOLON) {
		p.nextToken()
	}

	return stmt
}

func (p *Parser) parseBlockStatement() *ast.BlockStatement {
	block := &ast.BlockStatement{Token: p.curToken}
	block.Statements = []ast.Statement{}

	p.nextToken()

	for !p.curTokenIs(token.BRACE_CLOSE) && !p.curTokenIs(token.EOF) {
		stmt := p.parseStatement()
		if stmt != nil {
			block.Statements = append(block.Statements, stmt)
		}
		p.nextToken()
	}

	return block
}

// expressions

func (p *Parser) parseGroupedExpression() ast.Expression {
	p.nextToken()

	exp := p.parseExpression(LOWEST)

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	return exp
}

func (p *Parser) parseIndexExpression(left ast.Expression) ast.Expression {
	exp := &ast.IndexExpression{Token: p.curToken, Left: left}

	p.nextToken()
	exp.Index = p.parseExpression(LOWEST)

	if !p.expectPeek(token.BRACKET_CLOSE) {
		return nil
	}

	return exp
}

func (p *Parser) parseCallExpression(function ast.Expression) ast.Expression {
	exp := &ast.CallExpression{Token: p.curToken, Function: function}
	exp.Arguments = p.parseExpressionList(token.PAREN_CLOSE)

	return exp
}

func (p *Parser) parseIfExpression() ast.Expression {
	exp := &ast.IfExpression{Token: p.curToken}

	if !p.expectPeek(token.PAREN_OPEN) {
		return nil
	}

	p.nextToken()
	exp.Condition = p.parseExpression(LOWEST)

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	if !p.expectPeek(token.BRACE_OPEN) {
		return nil
	}

	exp.Consequence = p.parseBlockStatement()

	if p.peekTokenIs(token.ELSE) {
		p.nextToken()

		if !p.expectPeek(token.BRACE_OPEN) {
			return nil
		}

		exp.Alternative = p.parseBlockStatement()
	}

	return exp
}

func (p *Parser) parseExpression(precedence int) ast.Expression {
	parseUnary := p.unaryParseFns[p.curToken.Type]
	if parseUnary == nil {
		p.noUnaryParseFnError(p.curToken.Type)
		return nil
	}
	leftExp := parseUnary()

	for !p.peekTokenIs(token.SEMICOLON) && precedence < p.peekPrecedence() {
		parseBinary := p.binaryParseFns[p.peekToken.Type]
		if parseBinary == nil {
			return leftExp
		}

		p.nextToken()

		leftExp = parseBinary(leftExp)
	}

	return leftExp
}

func (p *Parser) parseUnaryExpression() ast.Expression {
	exp := &ast.UnaryExpression{
		Token:    p.curToken,
		Operator: p.curToken.Type,
	}

	p.nextToken()

	exp.Right = p.parseExpression(UNARY)

	return exp
}

func (p *Parser) parseBinaryExpression(left ast.Expression) ast.Expression {
	exp := &ast.BinaryExpression{
		Token:    p.curToken,
		Operator: p.curToken.Type,
		Left:     left,
	}

	precedence := p.curPrecedence()
	p.nextToken()
	exp.Right = p.parseExpression(precedence)

	return exp
}

// literals

func (p *Parser) parseFunctionLiteral() ast.Expression {
	lit := &ast.FunctionLiteral{Token: p.curToken}

	if !p.expectPeek(token.PAREN_OPEN) {
		return nil
	}

	lit.Parameters = p.parseFunctionParameters()

	if !p.expectPeek(token.BRACE_OPEN) {
		return nil
	}

	lit.Body = p.parseBlockStatement()

	return lit
}

func (p *Parser) parseFunctionParameters() []*ast.Identifier {
	identifiers := []*ast.Identifier{}

	if p.peekTokenIs(token.PAREN_CLOSE) {
		p.nextToken()
		return identifiers
	}

	p.nextToken()

	ident := &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
	identifiers = append(identifiers, ident)

	for p.peekTokenIs(token.COMMA) {
		p.nextToken()
		p.nextToken()
		ident := &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
		identifiers = append(identifiers, ident)
	}

	if !p.expectPeek(token.PAREN_CLOSE) {
		return nil
	}

	return identifiers
}

func (p *Parser) parseIdentifier() ast.Expression {
	return &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
}

func (p *Parser) parseBoolean() ast.Expression {
	return &ast.BooleanLiteral{Token: p.curToken, Value: p.curTokenIs(token.TRUE)}
}

func (p *Parser) parseIntegerLiteral() ast.Expression {
	lit := &ast.IntegerLiteral{Token: p.curToken}

	val, err := strconv.ParseInt(p.curToken.Literal, 0, 64)

	if err != nil {
		msg := fmt.Sprintf(
			"%v could not parse %q as integer",
			p.curToken.Pos,
			p.curToken.Literal,
		)
		p.errors = append(p.errors, msg)
		return nil
	}

	lit.Value = val

	return lit
}

func (p *Parser) parseStringLiteral() ast.Expression {
	return &ast.StringLiteral{Token: p.curToken, Value: p.curToken.Literal}
}

func (p *Parser) parseArrayLiteral() ast.Expression {
	array := &ast.ArrayLiteral{Token: p.curToken}

	array.Elements = p.parseExpressionList(token.BRACKET_CLOSE)

	return array
}

func (p *Parser) parseExpressionList(end token.TokenType) []ast.Expression {
	list := []ast.Expression{}

	if p.peekTokenIs(end) {
		p.nextToken()
		return list
	}

	p.nextToken()
	list = append(list, p.parseExpression(LOWEST))

	for p.peekTokenIs(token.COMMA) {
		p.nextToken()
		p.nextToken()
		list = append(list, p.parseExpression(LOWEST))
	}

	if !p.expectPeek(end) {
		return nil
	}

	return list
}

func (p *Parser) parseHashLiteral() ast.Expression {
	hash := &ast.HashLiteral{Token: p.curToken}

	hash.Pairs = make(map[ast.Expression]ast.Expression)

	for !p.peekTokenIs(token.BRACE_CLOSE) {
		p.nextToken()
		key := p.parseExpression(LOWEST)

		if !p.expectPeek(token.COLON) {
			return nil
		}

		p.nextToken()
		value := p.parseExpression(LOWEST)

		hash.Pairs[key] = value

		if !p.peekTokenIs(token.BRACE_CLOSE) && !p.expectPeek(token.COMMA) {
			return nil
		}
	}

	if !p.expectPeek(token.BRACE_CLOSE) {
		return nil
	}

	return hash
}

// helpers

func (p *Parser) curTokenIs(tt token.TokenType) bool {
	return p.curToken.Type == tt
}

func (p *Parser) peekTokenIs(tt token.TokenType) bool {
	return p.peekToken.Type == tt
}

func (p *Parser) expectPeek(tt token.TokenType) bool {
	if p.peekTokenIs(tt) {
		p.nextToken()
		return true
	} else {
		p.peekError(tt)
		return false
	}
}

const (
	_ int = iota
	LOWEST
	EQUALITY
	COMPARISON
	SUM
	PRODUCT
	UNARY
	CALL
	INDEX
)

var precedences = map[token.TokenType]int{
	token.EQUALS:       EQUALITY,
	token.NOT_EQUALS:   EQUALITY,
	token.LESS_THAN:    COMPARISON,
	token.GREATER_THAN: COMPARISON,
	token.PLUS:         SUM,
	token.MINUS:        SUM,
	token.ASTERISK:     PRODUCT,
	token.SLASH:        PRODUCT,
	token.PAREN_OPEN:   CALL,
	token.BRACKET_OPEN: INDEX,
}

func (p *Parser) peekPrecedence() int {
	if p, ok := precedences[p.peekToken.Type]; ok {
		return p
	}

	return LOWEST
}

func (p *Parser) curPrecedence() int {
	if p, ok := precedences[p.curToken.Type]; ok {
		return p
	}

	return LOWEST
}

// errors

func (p *Parser) Errors() []string {
	return p.errors
}

func (p *Parser) peekError(t token.TokenType) {
	msg := fmt.Sprintf(
		"%v expected %s, got %s",
		p.curToken.Pos,
		t,
		p.peekToken.Type,
	)
	p.errors = append(p.errors, msg)
}

func (p *Parser) noUnaryParseFnError(t token.TokenType) {
	msg := fmt.Sprintf(
		"%v no unary parse function for %s found",
		p.curToken.Pos,
		t,
	)
	p.errors = append(p.errors, msg)
}

File: src/parser/parser.zig
const std = @import("std");
const lexer = @import("../lexer/lexer.zig");
const token = @import("../token/token.zig");
const ast = @import("../ast/ast.zig");

pub const Parser = struct {
    lexer: *lexer.Lexer,
    errors: std.ArrayList([]const u8),

    cur_token: token.Token = undefined,
    peek_token: token.Token = undefined,

    allocator: std.mem.Allocator,

    const Self = @This();

    pub fn init(allocator: std.mem.Allocator, l: *lexer.Lexer) Parser {
        var parser = Parser{
            .lexer = l,
            .errors = std.ArrayList([]const u8).init(allocator),
            .allocator = allocator,
        };

        parser.nextToken();
        parser.nextToken();

        return parser;
    }

    pub fn parseProgram(self: *Self) !ast.Node {
        const program = try ast.Program.init(self.allocator);
        var statements = std.ArrayList(ast.Node).init(self.allocator);

        while (!self.curTokenIs(.eof)) {
            const stmt = try self.parseStatement();
            try statements.append(stmt);
            self.nextToken();
        }

        program.statements = statements.items;

        return .{ .program = program };
    }

    // statements

    fn parseStatement(self: *Self) !ast.Node {
        return switch (self.cur_token.type) {
            .@"return" => try self.parseReturnStatement(),
            .declaration => try self.parseDeclarationStatement(),
            else => try self.parseExpressionStatement(),
        };
    }

    fn parseReturnStatement(self: *Self) !ast.Node {
        const stmt = try ast.Return.init(self.allocator);
        stmt.token = self.cur_token;

        self.nextToken();

        stmt.return_value = try self.parseExpression(.lowest);

        while (self.peekTokenIs(.semicolon)) {
            self.nextToken();
        }

        return .{ .@"return" = stmt };
    }

    fn parseDeclarationStatement(self: *Self) !ast.Node {
        const stmt = try ast.Declaration.init(self.allocator);
        stmt.token = self.cur_token;

        try self.expectPeek(.identifier);

        const identifier = try ast.Identifier.init(self.allocator);
        identifier.token = self.cur_token;
        identifier.value = self.cur_token.literal;

        stmt.name = identifier;

        try self.expectPeek(.assign);

        self.nextToken();

        stmt.value = try self.parseExpression(.lowest);

        while (self.peekTokenIs(.semicolon)) {
            self.nextToken();
        }

        return .{ .declaration = stmt };
    }

    fn parseExpressionStatement(self: *Self) !ast.Node {
        const node = self.parseExpression(.lowest);

        while (self.peekTokenIs(.semicolon)) {
            self.nextToken();
        }

        return node;
    }

    fn parseBlockStatement(self: *Self) !ast.Node {
        const block = try ast.Block.init(self.allocator);
        var statements = std.ArrayList(ast.Node).init(self.allocator);

        self.nextToken();

        while (!self.curTokenIs(.brace_close) and !self.curTokenIs(.eof)) {
            const stmt = try self.parseStatement();
            try statements.append(stmt);
            self.nextToken();
        }

        block.statements = statements.items;

        return .{ .block = block };
    }

    // expressions

    fn parseExpression(self: *Self, precedence: Precedence) Error!ast.Node {
        var expr: ast.Node = switch (self.cur_token.type) {
            .paren_open => try self.parseGroupedExpression(),
            .@"if" => try self.parseIfExpression(),
            .bang => try self.parseUnaryExpression(),
            .minus => try self.parseUnaryExpression(),
            .function => try self.parseFunctionLiteral(),
            .identifier => try self.parseIdentifier(),
            .true => try self.parseBooleanLiteral(),
            .false => try self.parseBooleanLiteral(),
            .integer => try self.parseIntegerLiteral(),
            .string => try self.parseStringLiteral(),
            .bracket_open => try self.parseArrayLiteral(),
            .brace_open => try self.parseHashLiteral(),
            else => {
                self.noUnaryParseFnError(self.cur_token);
                return Error.NoUnaryParseFn;
            },
        };

        while (!self.peekTokenIs(.semicolon) and @intFromEnum(precedence) < @intFromEnum(self.peekPrecedence())) {
            const tok_type = self.peek_token.type;
            self.nextToken();
            expr = switch (tok_type) {
                .bracket_open => try self.parseIndexExpression(expr),
                .paren_open => try self.parseCallExpression(expr),
                .plus => try self.parseBinaryExpression(expr),
                .minus => try self.parseBinaryExpression(expr),
                .slash => try self.parseBinaryExpression(expr),
                .asterisk => try self.parseBinaryExpression(expr),
                .equals => try self.parseBinaryExpression(expr),
                .not_equals => try self.parseBinaryExpression(expr),
                .less_than => try self.parseBinaryExpression(expr),
                .greater_than => try self.parseBinaryExpression(expr),
                else => return expr,
            };
        }

        return expr;
    }

    fn parseGroupedExpression(self: *Self) Error!ast.Node {
        self.nextToken();

        const node = try self.parseExpression(.lowest);

        try self.expectPeek(.paren_close);

        return node;
    }

    fn parseIndexExpression(self: *Self, left: ast.Node) !ast.Node {
        const expr = try ast.Index.init(self.allocator);
        expr.token = self.cur_token;
        expr.left = left;

        self.nextToken();
        expr.index = try self.parseExpression(.lowest);

        try self.expectPeek(.bracket_close);

        return .{ .index = expr };
    }

    fn parseCallExpression(self: *Self, function: ast.Node) !ast.Node {
        const expr = try ast.Call.init(self.allocator);
        expr.token = self.cur_token;
        expr.function = function;
        expr.arguments = try self.parseExpressionList(.paren_close);

        return .{ .call = expr };
    }

    fn parseIfExpression(self: *Self) !ast.Node {
        const expr = try ast.If.init(self.allocator);
        expr.token = self.cur_token;

        try self.expectPeek(.paren_open);

        self.nextToken();
        expr.condition = try self.parseExpression(.lowest);

        try self.expectPeek(.paren_close);

        try self.expectPeek(.brace_open);

        expr.consequence = (try self.parseBlockStatement()).block;

        if (self.peekTokenIs(.@"else")) {
            self.nextToken();

            try self.expectPeek(.brace_open);

            expr.alternative = (try self.parseBlockStatement()).block;
        } else {
            expr.alternative = null;
        }

        return .{ .@"if" = expr };
    }

    fn parseUnaryExpression(self: *Self) !ast.Node {
        const expr = try ast.UnaryOperation.init(self.allocator);
        expr.token = self.cur_token;
        expr.operator = self.cur_token;

        self.nextToken();

        expr.right = try self.parseExpression(.unary);

        return .{ .unary_operation = expr };
    }

    fn parseBinaryExpression(self: *Self, left: ast.Node) !ast.Node {
        const expr = try ast.BinaryOperation.init(self.allocator);
        expr.token = self.cur_token;
        expr.operator = self.cur_token;
        expr.left = left;

        const precedence = self.curPrecedence();
        self.nextToken();
        expr.right = try self.parseExpression(precedence);

        return .{ .binary_operation = expr };
    }

    fn parseExpressionList(self: *Self, end: token.TokenType) ![]ast.Node {
        var list = std.ArrayList(ast.Node).init(self.allocator);

        if (self.peekTokenIs(end)) {
            self.nextToken();
            return list.items;
        }

        self.nextToken();
        try list.append(try self.parseExpression(.lowest));

        while (self.peekTokenIs(.comma)) {
            self.nextToken();
            self.nextToken();
            try list.append(try self.parseExpression(.lowest));
        }

        try self.expectPeek(end);

        return list.items;
    }

    // literals

    fn parseIdentifier(self: *Self) !ast.Node {
        const identifier = try ast.Identifier.init(self.allocator);
        identifier.token = self.cur_token;
        identifier.value = self.cur_token.literal;
        return .{ .identifier = identifier };
    }

    fn parseBooleanLiteral(self: *Self) !ast.Node {
        const boolean = try ast.BooleanLiteral.init(self.allocator);
        boolean.token = self.cur_token;
        boolean.value = self.curTokenIs(token.TokenType.true);
        return .{ .boolean_literal = boolean };
    }

    fn parseIntegerLiteral(self: *Self) !ast.Node {
        const integer = try ast.IntegerLiteral.init(self.allocator);
        integer.token = self.cur_token;
        integer.value = try std.fmt.parseInt(i32, self.cur_token.literal, 10);
        return .{ .integer_literal = integer };
    }

    fn parseStringLiteral(self: *Self) !ast.Node {
        const string = try ast.StringLiteral.init(self.allocator);
        string.token = self.cur_token;
        string.value = self.cur_token.literal;
        return .{ .string_literal = string };
    }

    fn parseArrayLiteral(self: *Self) !ast.Node {
        const array = try ast.ArrayLiteral.init(self.allocator);
        array.token = self.cur_token;
        array.elements = try self.parseExpressionList(.bracket_close);
        return .{ .array_literal = array };
    }

    fn parseHashLiteral(self: *Self) !ast.Node {
        const hash = try ast.HashLiteral.init(self.allocator);
        hash.token = self.cur_token;
        var pairs = std.AutoHashMap(ast.Node, ast.Node).init(self.allocator);

        while (!self.peekTokenIs(.brace_close)) {
            self.nextToken();
            const key = try self.parseExpression(.lowest);

            try self.expectPeek(.colon);

            self.nextToken();
            const value = try self.parseExpression(.lowest);

            try pairs.put(key, value);

            if (!self.peekTokenIs(.brace_close)) {
                try self.expectPeek(.comma);
            }
        }

        try self.expectPeek(.brace_close);

        hash.pairs = pairs;
        return .{ .hash_literal = hash };
    }

    fn parseFunctionLiteral(self: *Self) !ast.Node {
        const function = try ast.FunctionLiteral.init(self.allocator);
        function.token = self.cur_token;

        try self.expectPeek(.paren_open);

        function.parameters = try self.parseFunctionParameters();

        try self.expectPeek(.brace_open);

        function.body = (try self.parseBlockStatement()).block;

        return .{ .function_literal = function };
    }

    fn parseFunctionParameters(self: *Self) ![]*ast.Identifier {
        var identifiers = std.ArrayList(*ast.Identifier).init(self.allocator);

        if (self.peekTokenIs(.paren_close)) {
            self.nextToken();
            return identifiers.items;
        }

        self.nextToken();

        const identifier = try ast.Identifier.init(self.allocator);
        identifier.token = self.cur_token;
        identifier.value = self.cur_token.literal;
        try identifiers.append(identifier);

        while (self.peekTokenIs(.comma)) {
            self.nextToken();
            self.nextToken();

            const ident = try ast.Identifier.init(self.allocator);
            ident.token = self.cur_token;
            ident.value = self.cur_token.literal;
            try identifiers.append(ident);
        }

        try self.expectPeek(.paren_close);

        return identifiers.items;
    }

    // helpers

    fn nextToken(self: *Self) void {
        self.cur_token = self.peek_token;
        self.peek_token = self.lexer.nextToken();
    }

    fn curTokenIs(self: *Self, tok_type: token.TokenType) bool {
        return self.cur_token.type == tok_type;
    }

    fn peekTokenIs(self: *Self, tok_type: token.TokenType) bool {
        return self.peek_token.type == tok_type;
    }

    fn expectPeek(self: *Self, tok_type: token.TokenType) !void {
        if (self.peekTokenIs(tok_type)) {
            self.nextToken();
        } else {
            self.expectPeekError(tok_type);
            return Error.UnexpectedPeekToken;
        }
    }

    const Precedence = enum(u32) {
        lowest,
        equality,
        comparison,
        sum,
        product,
        unary,
        call,
        index,
    };

    fn getPrecedence(tok_type: token.TokenType) Precedence {
        return switch (tok_type) {
            .equals => .equality,
            .not_equals => .equality,
            .less_than => .comparison,
            .greater_than => .comparison,
            .plus => .sum,
            .minus => .sum,
            .asterisk => .product,
            .slash => .product,
            .paren_open => .call,
            .bracket_open => .index,
            else => .lowest,
        };
    }

    fn peekPrecedence(self: *Self) Precedence {
        return getPrecedence(self.peek_token.type);
    }

    fn curPrecedence(self: *Self) Precedence {
        return getPrecedence(self.cur_token.type);
    }

    // errors

    pub const Error = error{
        UnexpectedPeekToken,
        NoUnaryParseFn,
        OutOfMemory,
        Overflow,
        InvalidCharacter,
    };

    fn newError(self: *Self, message: []const u8, error_name: []const u8, tok: token.Token) void {
        var lines = std.mem.splitScalar(u8, self.lexer.input, '\n');
        var line: ?[]const u8 = null;
        var i: u32 = 1;
        while (lines.next()) |ln| : (i += 1) {
            if (i == tok.pos.ln) {
                line = ln;
                break;
            }
        }

        var pointer = std.ArrayList(u8).init(self.allocator);
        for (0..(tok.pos.col - 1)) |_| {
            _ = pointer.append(' ') catch null;
        }

        var pointer_width = tok.literal.len;

        if (pointer_width == 0) {
            pointer_width = 1;
        }

        if (tok.type == .string) {
            pointer_width += 2;
        }

        for (0..(pointer_width)) |_| {
            _ = pointer.append('^') catch null;
        }
        _ = pointer.appendSlice(" here") catch null;

        const msg = std.fmt.allocPrint(
            self.allocator,
            "{s}: {s}\n{s}\n{s}\n",
            .{
                tok.pos.string(self.allocator),
                message,
                line orelse "???",
                pointer.items,
            },
        ) catch |err| {
            std.debug.print("{s} {s}", .{ error_name, @errorName(err) });
            return;
        };
        self.errors.append(msg) catch |err| std.debug.print("{s} {s}", .{ error_name, @errorName(err) });
    }

    fn expectPeekError(self: *Self, tok_type: token.TokenType) void {
        const msg = std.fmt.allocPrint(
            self.allocator,
            "expected {s}, got {s}",
            .{
                tok_type.string(self.allocator),
                self.peek_token.string(self.allocator),
            },
        ) catch |err| {
            std.debug.print("expectPeekError {s}", .{@errorName(err)});
            return;
        };
        self.newError(msg, "expectPeekError", self.peek_token);
    }

    fn noUnaryParseFnError(self: *Self, tok: token.Token) void {
        const msg = std.fmt.allocPrint(
            self.allocator,
            "no unary parse function for {s} found",
            .{
                tok.string(self.allocator),
            },
        ) catch |err| {
            std.debug.print("noUnaryParseFnError {s}", .{@errorName(err)});
            return;
        };
        self.newError(msg, "noUnaryParseFnError", tok);
    }
};

const parser_test = @import("./parser_test.zig");

test {
    parser_test.hack();
}
